File: examples/lh-server/typescript/src/21-structdefs/main.ts
import {
  buildPutStructDefRequest,
  createTaskWorker,
  LHConfig,
  Workflow,
} from "littlehorse-client";
import { z } from "zod";
import { closeOnShutdown } from "../config.js";
import { Car } from "./schemas.js";

function describeCar(car: z.infer<typeof Car>): string {
  return `You drive a ${car.make} ${car.model}`;
}

const config = LHConfig.from({});
const client = config.getClient();
const worker = createTaskWorker(describeCar, "describe-car", config, {
  inputVars: { car: Car },
  outputSchema: z.string(),
});

await client.putStructDef(buildPutStructDefRequest(Car));

const workflow = Workflow.newWorkflow("quickstart", (wf) => {
  const inputCar = wf.declareStruct("input-car", Car).required();
  wf.execute("describe-car", inputCar);
});

await closeOnShutdown([worker]);
await worker.registerTaskDef();
await workflow.registerWfSpec(config);
console.log("Run with: npm run structdefs:run");
await worker.start();

File: examples/lh-server/python/21-structdefs/main.py
import asyncio
import sys

import littlehorse
from littlehorse.config import LHConfig
from littlehorse.worker import LHTaskWorker
from tasks import Car, describe_car
from workflow import quickstart_wf


def register(config: LHConfig) -> None:
    littlehorse.create_struct_def(Car, config)
    littlehorse.create_task_def(describe_car, "describe-car", config)
    littlehorse.create_workflow_spec(quickstart_wf(), config)


async def workers(config: LHConfig) -> None:
    await littlehorse.start(LHTaskWorker(describe_car, "describe-car", config))


if __name__ == "__main__":
    config = LHConfig()
    if len(sys.argv) != 2 or sys.argv[1] not in {"register", "workers"}:
        raise SystemExit("Usage: main.py [register|workers]")
    register(config) if sys.argv[1] == "register" else asyncio.run(workers(config))
